### PR TITLE
Coords after

### PR DESCRIPTION
**Pull request by fgrunewald**
This is essentially an updated version of the coordinate builder but after the fact. This fixes mutations and everything where at the CG level coordinates are missing. This could also fix modifications like phosphorylation that are not actually present in the mapped pdb. Currently missing is a good routine for rebuilding VS. Luckily there is a very well tested routine in polyply. For now it only rebuilds VSN but that's good enough for proteins in Martini3 our main target. 

@pckroon 
Should we move the vs_builder from polyply to here eventually?

@Tsjerk 
As we discussed can you update the doc-strings with a little more information regarding how the coordinates are built and what quality checks are done. Also we'll need some tests here. Perhaps @pckroon can help with those as well a little. 

@paulocts 
You can use this branch for checking some examples, if you like.

### CODE REVIEW

File: vermouth/processors/generate_coordinates.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+        if all(np.isnan(molecule.nodes[ndx]['position'])):
+            missing.append(ndx)
+    return missing
+
+def get_anchors(molecule, indices):
+    '''
+    Determine particles with known coordinates connected to
+    particles with given indices.
+    '''
+    anchors = []
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:44]: start with anchors as a set and add to that, rather first building a list and turning it to a set in the end

File: vermouth/processors/generate_coordinates.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+
+def double_cone(n=24, distance=0.125, angle=109.4, axis=None):
+    '''Create positions for two consecutive (SP3) bonded particles'''
+    P = conicring(n, distance, angle, axis)
+    S = np.array([P @ align_z(x) + x for x in P])
+    return P, S
+
+
+def triple_cone(n=24, distance=0.125, angle=109.4):
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:91]: I think we briefly discussed this earlier, but there's no easy way of generalizing the double/triple/quadrupal cone constructions right? Even as a recursive function?

File: vermouth/processors/generate_coordinates.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    Returns
+    -------
+        None
+    '''
+    a = molecule.nodes[anchor]['position']
+    if control == None:
+        B = [
+            molecule.nodes[ix]['position']
+            for ix, v in molecule[anchor].items()
+            if v != {}
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:125]:
```suggestion
            if not v
```
?

File: vermouth/processors/generate_coordinates.py [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    `down`, referring to the direction with respect to the
+    cross-product of the two substituents. If these control
+    atoms are not given explicitly in order, then the control
+    atoms are inferred from the connections of the anchor with
+    known coordinates. In that case, the specific chirality
+    of the center (R or S) cannot be controlled.
+
+    NOTE: This function should probably be able to determine
+          the weights of the substituents to allow placement
+          based on R or S chirality.
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:156]:
```suggestion
    TODO: This function should probably be able to determine
          the weights of the substituents to allow placement
          based on R or S chirality.
```

File: vermouth/processors/generate_coordinates.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    Returns
+    -------
+        None
+    '''
+    a = molecule.nodes[anchor]['position']
+    if control is None:
+        B = [
+            molecule.nodes[ix]['position']
+            for ix, v in molecule[anchor].items()
+            if v != {}
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:171]:
```suggestion
            if not v
```

File: vermouth/processors/generate_coordinates.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        self.bases = limbs[0][1]
+        self.limbs = [ s[2] for s in limbs ]
+
+    def __str__(self):
+        return ':'.join([str(self.anchors), str(self.bases), str(self.limbs)])
+
+    def extrude(self, distance=0.125, coords=None, contact=0.5):
+        mol = self.molecule
+
+        for i, (a, b) in enumerate(zip(self.anchors, self.bases)):
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:210]:
```suggestion
        for idx, (anchor, base) in enumerate(zip(self.anchors, self.bases)):
```
It costs nothing to expand these variable names, and it makes the code *so much* clearer...

File: vermouth/processors/generate_coordinates.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+                # Trivial chiral - except for the actual chirality
+                # a simple 'up' or 'down' flag would be helpful
+                # amino acid side chain is 'up'
+                up, down = _chiral(mol, a, target[0], target[1], control=b)
+                # Simply update this segment and return it
+                for k in self.limbs:
+                    k.discard(target[0])
+                self.bases[i] = [a]
+                self.anchors[i] = target[0]
+                return [self]
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:234]: Deduplicate this with the other branches

File: vermouth/processors/generate_coordinates.py [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+                    #                        self.molecule,
+                    #                        [(t, [[a]], g)],
+                    #                        self.coords,
+                    #                        self.contact
+                    #                    )
+                    #                )
+                    #  This commented-out stuff may be useful for speeding up
+                    #  coordinate generation, because it avoids needing lookup
+                    #  of missing coordinates in subsequent cycles.
+                    #return out
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:276]: This is missing a "but". Why isn't it used?

File: vermouth/processors/generate_coordinates.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+                for i, a in m.atoms
+            ]
+        )
+
+        for molecule in system.molecules:
+            self.run_molecule(molecule)
+
+    def run_molecule(self, molecule):
+        """
+        Process a single molecule. Must be implemented by subclasses.
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:335]: This is not the docstring I am looking for. Move along.

File: vermouth/processors/generate_coordinates.py [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+    A processor for generating coordinates for atoms without.
+    """
+    def __init__(self, contact=0.5):
+        super().__init__()
+        self.contact = contact
+        self.coordinates = None
+
+    def run_system(self, system):
+        """
+        Process `system`.
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:314]: This is not the docstring you are looking for.

File: vermouth/processors/generate_coordinates.py [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+        while prev - len(missing) > 0:
+            # Limbs: subgraphs of missing atoms with anchors and bases
+            limbs = []
+            for g in nx.connected_components(molecule.subgraph(missing)):
+                anchors = list(get_anchors(molecule, g))
+                bases = [ list(get_anchors(molecule, [a])) for a in anchors ]
+                limbs.append((anchors, bases, g))
+            limbs.sort()
+
+            # Segment: subgraphs of limbs sharing anchors and bases
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:359]: Some of these comments should be moved to the class

File: vermouth/processors/generate_coordinates.py [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+            for lim in limbs:
+                if lim[0] != prev:
+                    segments.append([])
+                    prev = lim[0]
+                segments[-1].append(lim)
+            segments = [ Segment(molecule, limbs) for limbs in segments ]
+
+            # Extrude segments one layer at a time
+            while segments:
+                s = segments.pop(0)
```
Comment (pckroon) [inline vermouth/processors/generate_coordinates.py:371]: Nit: Not ideal to be popping from the left all the time. Is it possible to reverse sort the limbs above, and pop from the right here?

**Review by pckroon — CHANGES_REQUESTED** [review]
I reviewed most of this code before, and I still don't like it very much. If this is merged as-is it'll incur a massive maintenance debt. If it breaks I'll have no idea where to look for the issue. The lack of comments and docstrings (the ones that are there are insufficient) acerbates the issue. Coupled to the absence of test this is time-bomb.
I assume I shouldn't get my hopes up regarding the missing tests, failing pylint, and docs?

I would very much welcome the VS routines from polyply here (make sure you can add the required citations, etc), but it will depend a little bit on the dependencies. Martinize2 only has a soft dependency on scipy, and making that a hard dependency needs to be a conscious decision.

**Comment by fgrunewald** [comment]
will take a second look at this in a clean PR that's not based on some other branches.

**Pull request closed without merging**